### ISSUE TITLE
Figure out deployment strategy for Bramble

### ISSUE THREAD

**Issue by humphd** [mozilla/thimble.webmaker.org#474]
We're ~1 week away from having the Brackets-in-Thimble (i.e., "Bramble") work be ready for people to start using.  You can already play with it on heroku at https://cdot-bramble.herokuapp.com/ (it's the code at https://github.com/sedge/thimble.webmaker.org/tree/mvpIntegration). Basically, it's a drop-in replacement for Thimble.

Obviously switching over to this fully isn't wise until it's been tested and we fix various things based on feedback.  As such, I want to come up with a strategy to support both Thimble as the default and Bramble as a thing you can opt into.  What do you think about something like:

* https://thimble.webmaker.org gives you default Thimble
* https://thimble.webmaker.org/?bramble=1 gives you Thimble running Bramble

In this way we could have both running and let users keep using Thimble while we still test and develop Bramble.

Thoughts?

cc @sedge, @simonwex, @jdotpz, @davidascher

**Comment by Pomax** [comment]
additional question: has the work on bramble given rise to a battery of tests that we can use going forward to prevent regression?

**Comment by humphd** [comment]
We inherit all of Brackets' test suite for free (as well as them maintaining them and not regressing) by using their code.

Do we have tests for the parts we've made?  Beyond linting, no, though the code surface is pretty small relatively.

**Comment by sedge** [comment]
It wouldn't be too hard to serve from a friendlycode directory with brackets or a friendlycode directory without it based on a switch. I have a pretty clear idea of how to accomplish this.

**Comment by Pomax** [comment]
the less remains of original friendlycode, the easier it'll be to start adding tests in for those parts that remain - Slowparse is backed by server side tests, the rest of friendlycode I think were qunit tests

**Comment by humphd** [comment]
From discussions with @jbuck today on irc, we're going to have to change the way we serve the Brackets-friendlycode iframe.  Right now, we have an architecture where Thimble hosts the Brackets-friendlycode iframe, which holds the Brackets editor and also a preview iframe.  The way that our new live dev works, when a page is previewed (i.e., `/index.html`), we parse and rewrite it in order to deal with any external resources (.js or .css files in the browser filesystem), converting these to BLOB URL objects in memory.  At this point, the page itself is also converted to a BLOB URL in memory, and we do `iframe.src = blobURL;`, which means the iframe ends up hosting a URL of something like `blob:https%3A//thimble.webmaker.org/<some-uuid>`.

According to the CSP spec this is no good living in the same domain as the main Thimble app, see http://www.w3.org/TR/CSP2/#source-list-guid-matching, specifically:

<blockquote>policy authors should be aware that allowing "data:" URLs is equivalent to unsafe-inline and allowing "blob:" or "filesystem:" URLs is equivalent to unsafe-eval."</blockquote>

In other words, it's trivial to do bad things with script that affect the Thimble domain.  The answer is to split the loading of Bramble into two parts:
* Thimble app from https://thimble.webmaker.org, which holds an iframe for the editor and preview
* Editor and Preview iframe loads from a second domain (probably mozillathimblelivepreview.net, since we already own/use that for this purpose).  Jon also suggests that we could even get Travis CI to do deploys to that domain, so we don't need anyone to upload content manually

I'm a bit unclear how to do this without breaking current Thimble, and I think we want to run both in parallel while we test and fix things in Bramble.  I'm going to task @sedge with doing things necessary for this on our side, and I'll ask that you all give him help as needed in order to properly guide our progress.

I'll ask more questions here or on irc as they become important.

**Comment by humphd** [comment]
Some of the work for the above is happening in https://github.com/humphd/brackets/issues/107 if you want to follow along.

Also, I have a question about how we're going to deploy in the short/medium-term.  Is it possible for us to land our current work for Bramble in mozilla/thimble.webmaker.org on a "bramble" branch, and maintain that branch for a while alongside master?  My goal is to be able to have thimble.webmaker.org and maybe bramble.webmaker.org or something both be deployed and have all the automation stuff figured out.  Then we can continue to run Thimble, but properly use and test Bramble while we iron out issues.

What has to happen for that, and who needs to be involved?

**Comment by jbuck** [comment]
I think the quickest and easiest way to get this rolling is:

1) Create a bramble branch on this repo 
2) Create a Heroku app from the bramble branch that hooks up to everything else on staging. Configure Travis CI to push the the bramble branch to this Heroku app
3) Create `https://bramble.mofostaging.net`
4) Iterate on Bramble until it's ready

Thoughts?

**Comment by humphd** [comment]
Sounds good to me, especially since it parallels what we've already done.  @sedge, what do you think?

**Comment by sedge** [comment]
The last piece of the puzzle is to get `make-valet` working with it. @jbuck Does mofo have a domain we can use for implementing that?

**Comment by jbuck** [comment]
No need, we'll just use the current staging make-valet instead of running a separate one

**Comment by jbuck** [comment]
Right now I'm stuck on configuring Cloudfront correctly. I'm currently receiving a 502 Bad Gateway when the Cloudfront Origin is set to bramble-staging.herokuapp.com.

@jdotpz any ideas?

**Comment by sedge** [comment]
@jbuck Well that's your problem right there! Point it at `cdot-bramble.herokuapp.com`

**Comment by jdotpz** [comment]
I hypothesize it's due to the SSL cert.  You get different errors based on protocol on the client, and I see this in docs:

"If the origin server returns an invalid certificate or a self-signed certificate, or if the origin server returns the certificate chain in the wrong order, CloudFront drops the TCP connection, returns HTTP error code 502, and sets the X-Cache header to Error from cloudfront."

I'm betting it's invalid cert coming across for the host headers it's sending.

**Comment by humphd** [comment]
We did the work to move the editor+preview iframes into a new iframe so we could lock down Thimble and isolate user-defined JS, but our plan to host it from github.io is not going to work, because of mixed content.  Github's HTTPS on github.io is a 301 to your site via http :(

We're going to need a new solution here for where to put this static iframe source until we have it in S3.  Thoughts?

**Comment by sedge** [comment]
Could we host it on thimblepreviewloader.net or whatever it is?

**Comment by Pomax** [comment]
I don't see why not. give it its own name (`bracketspreviewloader.html`?) and then it can just live alongside the thimble one

**Comment by sedge** [comment]
It's more than one file, so it would be a static directory: `/bramble/dist`

**Comment by Pomax** [comment]
how many files would this be? We already have that bucket so if they're all static files, I don't see why we couldn't put that on S3. @jbuck thoughts?

**Comment by jbuck** [comment]
Yeah, we can upload everything to S3

**Comment by jbuck** [comment]
Okay, I finally got https://bramble.mofostaging.net working. The key was to not have Cloudfront forward the Host header to Heroku!

**Comment by jdotpz** [comment]
Definitely cheaper than $20. :)

**Comment by sedge** [comment]
@jbuck Can you add me as a collaborator on the heroku instance you set up? Also, I'll need the keys for the previewloader domain's S3 bucket to get auto-deploying up and running.

**Comment by jbuck** [comment]
@sedge I've PMed you AWS S3 credentials. What's your email for Heroku?

**Comment by sedge** [comment]
**Update:** Our brackets fork (**bramble**) is now automatically deploying to thimble's preview s3 bucket and can be accessed here: https://mozillathimblelivepreview.net/bramble/dist/

The next step is to merge our work on super-thimble into thimble's `bramble` branch and set up automatic deployment to heroku. @jbuck Can I get your go ahead?

**Comment by jbuck** [comment]
@sedge yeah go for it

**Comment by sedge** [comment]
[bramble.mofostaging.net](https://bramble.mofostaging.net) is live! 

To update the editor, merge JSLinted changes into `humphd/brackets#bramble`.
To update thimble, merge JSLinted changes into `mozilla/thimble.webmaker.org#bramble`.

Automagically, things will "just work"! 

Any final comments before I close the issue?

**Comment by simonwex** [comment]
This is awesome. :+1:

**Comment by humphd** [comment]
Indeed, well done.

![Indeed](http://i.giphy.com/jShr8wkP38XTO.gif)

**Comment by jbuck** [comment]
fin
